hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+instance Functor (Utxo' era) where
+  fmap fn (Utxo u) = Utxo (fmap fn u)
 
-type CardanoAuxiliaryData = StrictMaybe (AuxiliaryData CardanoEra)
+instance Foldable (Utxo' era) where
+  foldMap fn = foldMap fn . unUtxo
+  foldr fn zero = foldr fn zero . unUtxo
 
-type CardanoAddress = Cardano.Addr StandardCrypto
+instance Semigroup Utxo where
+  Utxo uL <> Utxo uR = Utxo (uL <> uR)
 
-type CardanoKeyPair = Cardano.KeyPair 'Cardano.Payment StandardCrypto
+instance Monoid Utxo where
+  mempty = Utxo mempty
 
-type TxIn = Cardano.TxIn StandardCrypto
+instance ToJSON Utxo where
+  toJSON = toJSON . unUtxo
 
-type TxOut = Cardano.TxOut CardanoEra
+instance FromJSON Utxo where
+  parseJSON = fmap Utxo . parseJSON
 
-signWith ::
-  Cardano.TxId StandardCrypto ->
-  Cardano.KeyPair 'Cardano.Payment StandardCrypto ->
-  Cardano.WitVKey 'Cardano.Witness StandardCrypto
-signWith (Cardano.TxId safeHash) credentials =
-  Cardano.WitVKey
-    (asWitness $ Cardano.vKey credentials)
-    (signedDSIGN @StandardCrypto (Cardano.sKey credentials) (extractHash safeHash))
+instance Arbitrary Utxo where
+  -- TODO: shrinker!
+  arbitrary = genUtxo
 
---
--- Balance
---
+prettyUtxo :: (TxIn, TxOut ctx era) -> Text
+prettyUtxo (k, TxOut _ (txOutValueToValue -> v) _) =
+  T.drop 54 (renderTxIn k) <> " ↦ " <> prettyValue v
 
-prettyBalance :: Balance tx -> Text
-prettyBalance Balance{lovelace, assets} =
-  let (ada, decimal) = lovelace `quotRem` 1000000
-      padLeft c n str = T.takeEnd n (T.replicate n c <> str)
-   in unwords $
-        [ show ada <> "." <> padLeft "0" 6 (show decimal)
-        , "₳"
+-- FIXME: This function is wrong! It's mapping a transaction's own inputs to its
+-- own outputs. Whoops. It's currently used in Hydra.Chain.Direct.Tx where
+-- the calling code only look at the outputs of the transactions and also in the
+-- generators of the local-cluster (Whoops bis).
+utxoFromTx :: CardanoTx -> Utxo
+utxoFromTx (Tx body@(ShelleyTxBody _ ledgerBody _ _ _ _) _) =
+  let txOuts = toList $ Ledger.Alonzo.outputs' ledgerBody
+      txIns =
+        [ Ledger.TxIn (toLedgerTxId $ getTxId body) ix
+        | ix <- [0 .. fromIntegral (length txOuts)]
         ]
-          ++ if null assets
-            then mempty
-            else
-              [ "and"
-              , show (Map.size assets)
-              , "asset(s)"
-              ]
+   in Utxo $ Map.fromList $ zip txIns txOuts
 
 --
---  Transaction Id
+-- Tx
 --
 
-instance Crypto crypto => ToJSON (Cardano.TxId crypto) where
-  toJSON = String . txIdToText @crypto
+instance IsTx (CardanoTx) where
+  type TxIdType (CardanoTx) = TxId
+  type UtxoType (CardanoTx) = Utxo
+  type ValueType (CardanoTx) = Value
 
-instance Crypto crypto => FromJSON (Cardano.TxId crypto) where
-  parseJSON = withText "base16 encoded TxId" txIdFromText
+  txId = getTxId . getTxBody
+  balance (Utxo u) =
+    let aggregate (Ledger.Alonzo.TxOut _ value _) = (<>) (fromMaryValue value)
+     in Map.foldr aggregate mempty u
```

Comment:
As you have a `Foldable` instance on `Utxo` (see comment above, but if we keep it, we should use it), you could use `foldMap` and leverage the `Monoid Value`. So this should work:

```suggestion
  balance = foldMap (\(Ledger.Alonzo.TxOut _ value _) -> fromMaryValue value)
```

Reply:
Indeed, I just took the code as it was, but can be simplified now :)